interface/modules/zend_modules/module/CodeTypes/src/CodeTypes/Listener/CodeTypeEventsSubscriber.php

```diff
@@ -0,0 +1,207 @@
+        ];
+    }
+
+    public function onSqlUpgradeEvent(SQLUpgradeEvent $event)
+    {
+        // grab our currently installed code types, check for SNOMED-CT AND CPT4 and then update accordingly.
+        // SELECT * FROM code_types WHERE ct_active=1 ORDER BY ct_seq, ct_key
+        $activatedCodeTypes = QueryUtils::fetchRecords("SELECT ct_key FROM code_types WHERE ct_active=1 ORDER BY ct_seq, ct_key");
+        if (empty($activatedCodeTypes)) {
+            return;
+        }
+        // we want to push out to the system that we are making changes...
+        $logger = function ($message) use ($event) {
+            // make sure we escape this here.
+            $event->getSqlUpgradeService()->flush_echo(text($message) . "<br />");
+        };
+
+        foreach ($activatedCodeTypes as $record) {
+            if ($record['ct_key'] == self::CODE_TYPE_CPT4) {
+                if ($this->shouldUpdateCPT4Mappings()) {
+                    $logger("Updating " . $record['ct_key'] . " Mappings");
+                    $this->updateCPT4Mappings($logger);
+                } else {
+                    $logger("Skipping Section #updateCPT4Mappings");
+                }
+            } else if ($this->isSnomedCodeType($record['ct_key'])) {
+                if ($this->shouldUpdateSNOMEDMappings()) {
+                    $logger("Updating " . $record['ct_key'] . " Mappings");
+                    $this->updateSNOMEDCTMappings($logger);
+                } else {
+                    $logger("Skipping Section #updateSNOMEDMappings type=" . $record['ct_key']);
+                }
+            }
+        }
+    }
+
+    public function onCodeTypeInstalledPreEvent(CodeTypeInstalledEvent $event)
+    {
+        error_log("Got here in pre installed event");
+    }
+
+    public function onCodeTypeInstalledEvent(CodeTypeInstalledEvent $event)
+    {
+        if ($event->getCodeType() == "SNOMED") {
+            // check if we have SNOMED codes installed and update our list options
+            $this->updateSNOMEDCTMappings();
+        } else if ($event->getCodeType() == "CPT4" && $this->shouldUpdateCPT4Mappings()) {
+            // check if we have CPT4 codes installed and update our list options
+            $this->updateCPT4Mappings();
+        }
+    }
+
+    private function isSnomedCodeType($codeType)
+    {
+        return in_array($codeType, [self::CODE_TYPE_SNOMED, self::CODE_TYPE_SNOMED_CT, self::CODE_TYPE_SNOMED_PR]);
+    }
+
+    private function exists_code_table($tableName)
+    {
+        // make sure our table is installed
+        $table_records = QueryUtils::fetchRecords("SHOW TABLES LIKE ?", [$tableName]);
+        if (empty($table_records)) {
+            (new SystemLogger())->debug("table for code_type does not exist", ['tableName' => $tableName]);
+        }
+        return !empty($table_records);
+    }
+
+    private function shouldUpdateCPT4Mappings()
+    {
+        if (!$this->exists_code_table('codes_cpt')) {
```

Comment:
I didn't realize there was a codes_cpt table. How does that table get created?

Reply:
@bradymiller It exists in the MU3 demos and I thought that is where the CPT4 codes are loaded from.  I don't exactly know how it gets imported as the standard_tables_capture.inc doesn't have a process for it.

Reply:
@adunsulag , @stephenwaite , i think most users import these into the codes table. Thoughts here? Should we check if each one exists in codes table and if it does, then ok to use the CPT code?

Reply:
I just checked my codes table with the MU3 database and it has identical CPT4 codes as far as I can tell.  I think I'll just change the code up to use the codes table.  I'll make the exists_code_table to be a is_active_code_type and just check against the code_type table.

I'm glad you caught this as I was just following the same logic as what we do with snomed's sct tables.

Reply:
I updated this to use the codes table.  I tested this on both the MU3 database and the demo database (openemr-cmd drid) and it looks like it is working correctly.

Reply:
nice! then can bring it on in :)